src/main/java/software/amazon/cloudwatchlogs/emf/model/MetricsContext.java

```diff
@@ -203,36 +203,72 @@ public MetricsContext createCopyWithContext() {
      * metrics in one log event. If there're more than 100 metrics, we split the metrics into
      * multiple log events.
      *
+     * <p>If a metric has more than 100 data points, we also split the metric.
+     *
      * @return the serialized strings.
      * @throws JsonProcessingException if there's any object that cannot be serialized
      */
     public List<String> serialize() throws JsonProcessingException {
-        if (rootNode.metrics().size() <= Constants.MAX_METRICS_PER_EVENT) {
+        if (rootNode.metrics().size() <= Constants.MAX_METRICS_PER_EVENT
+                && !anyMetricWithTooManyDataPoints(rootNode)) {
             return Arrays.asList(this.rootNode.serialize());
         } else {
             List<RootNode> nodes = new ArrayList<>();
             Map<String, MetricDefinition> metrics = new HashMap<>();
-            int count = 0;
-            for (MetricDefinition metric : rootNode.metrics().values()) {
-                metrics.put(metric.getName(), metric);
-                count++;
+            Queue<MetricDefinition> metricDefinitions =
+                    new LinkedList<>(rootNode.metrics().values());
```

Comment:
1. Personal preference here is to not use the abstract interface for your local variable. The concrete type (LinkedList) won't be changing and there's not much of a need to limit yourself to only the interface. This is not a perf comment (dynamic dispatch won't be necessary in this context) just a style comment so YMMV.
2. This constructor forces [enumeration of the set and allocation of a new list](https://hg.openjdk.java.net/jdk8/jdk8/jdk/file/687fd7c7986d/src/share/classes/java/util/LinkedList.java#l422). Is this really necessary? It shouldn't be a significant issue, just something to think about.